### PR TITLE
Add kubefed installation to E2E flow as an option

### PR DESCRIPTION
**Pull request by dimaunx**
This PR is adding kubefed=true command to E2E flow.

### CODE REVIEW

File: Dockerfile.dapper [hunk 1]
```diff
@@ -18,11 +19,13 @@ RUN apt-get -q update && \
...
-    cp linux-${ARCH}/helm /usr/bin/ && \
-    curl -Lo /usr/bin/kind https://github.com/kubernetes-sigs/kind/releases/download/v0.3.0/kind-linux-${ARCH} && chmod a+x /usr/bin/kind && \
+    curl "https://storage.googleapis.com/kubernetes-helm/helm-${HELM_VERSION}-linux-${ARCH}.tar.gz" | tar -xzf - && \
+    cp linux-${ARCH}/helm /usr/bin/ && chmod a+x /usr/bin/helm && \
+    curl -LO "https://github.com/kubernetes-sigs/kubefed/releases/download/v${KUBEFED_VERSION}/kubefedctl-${KUBEFED_VERSION}-linux-${ARCH}.tgz" && \
+    tar -xzf kubefedctl-${KUBEFED_VERSION}-linux-${ARCH}.tgz && cp kubefedctl /usr/bin/ && chmod a+x /usr/bin/kubefedctl && \
+    curl -Lo /usr/bin/kind "https://github.com/kubernetes-sigs/kind/releases/download/${KIND_VERSION}/kind-linux-${ARCH}" && chmod a+x /usr/bin/kind && \
     go get -v github.com/onsi/ginkgo/ginkgo && go get -v github.com/onsi/gomega/... && \
-    go get golang.org/x/tools/cmd/goimports
+    go get -v golang.org/x/tools/cmd/goimports
```
Comment (mangelajo) [inline Dockerfile.dapper]: Just a question out of curiosity (not related to this patch, and no action required):

is there any big advantage in doing it all in one big run with && .. && .. && 
beyond avoiding several layers? Is the layer overhead too big?

It's a little bit annoying to wait for the whole container being rebuilt for a simple change :) :-)

File: scripts/kind-e2e/e2e.sh [hunk 2]
```diff
@@ -190,6 +192,25 @@ function enable_logging() {
...
+function enable_kubefed() {
+    kubectl config use-context cluster1
+    if kubectl rollout status deploy/kubefed-controller-manager -n ${KUBEFED_NS} > /dev/null 2>&1; then
+        echo Kubefed already installed, skipping setup...
+    else
+        helm repo add kubefed-charts https://raw.githubusercontent.com/kubernetes-sigs/kubefed/master/charts
+        helm install kubefed-charts/kubefed --version=0.1.0-rc2 --name kubefed --namespace ${KUBEFED_NS} --set controllermanager.replicaCount=1 > /dev/null 2>&1
+        for i in 1 2 3; do
+            kubefedctl join cluster${i} --cluster-context cluster${i} --host-cluster-context cluster1 --v=2
+            #master_ip=$(docker inspect -f '{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' cluster${i}-control-plane | head -n 1)
```
Comment (mangelajo) [inline scripts/kind-e2e/e2e.sh]: Will we need those bits of commented code?, it'd be nice to clean up :)

File: scripts/kind-e2e/nginx-federated.sh [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+set -e
+
+TEST_NS="test-ns"
+export KUBECONFIG=$(echo $(git rev-parse --show-toplevel)/output/kind-config/local-dev/kind-config-cluster{1..3} | sed 's/ /:/g')
+kubectl config use-context cluster1
+
+echo "Creating a test namespace ${TEST_NS}."
+kubectl create ns ${TEST_NS}
+
+echo "Creating a ngnix application resources in namespace ${TEST_NS}."
```
Comment (mangelajo) [inline scripts/kind-e2e/nginx-federated.sh]: ngnix -> nginx

File: scripts/kind-e2e/nginx-federated.sh [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        volumeMounts:
+        - name: federated-web-file
+          mountPath: /usr/share/nginx/html/
+      volumes:
+        - name: federated-web-file
+          configMap:
+            name: federated-web-file
+EOF
+
+cat <<EOF | kubectl apply -n ${TEST_NS} -f -
```
Comment (mangelajo) [inline scripts/kind-e2e/nginx-federated.sh]: I believe you can also do:

kubectl apply -n ${TEST_NS} -f - <<EOF

(I haven't tried though..)

File: scripts/kind-e2e/nginx-federated.sh [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: federated-web-file
+data:
+  index.html: "Hello from Kubernetes Cluster Federation!"
+EOF
+
+cat <<EOF | kubectl apply -n ${TEST_NS} -f -
```
Comment (mangelajo) [inline scripts/kind-e2e/nginx-federated.sh]: kubectl apply -n ${TEST_NS} -f - <<EOF

File: scripts/kind-e2e/nginx-federated.sh [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+TEST_NS="test-ns"
+export KUBECONFIG=$(echo $(git rev-parse --show-toplevel)/output/kind-config/local-dev/kind-config-cluster{1..3} | sed 's/ /:/g')
+kubectl config use-context cluster1
+
+echo "Creating a test namespace ${TEST_NS}."
+kubectl create ns ${TEST_NS}
+
+echo "Creating a ngnix application resources in namespace ${TEST_NS}."
+cat <<EOF | kubectl apply -n ${TEST_NS} -f -
```
Comment (mangelajo) [inline scripts/kind-e2e/nginx-federated.sh]: kubectl apply -n ${TEST_NS} -f - <<EOF

File: scripts/kind-e2e/e2e.sh (hunk 2 above)
Comment (dimaunx) [reply scripts/kind-e2e/e2e.sh]: Want to keep this as a placeholder for future @skitt work. He will require it to make e2e to work without dapper.

File: Dockerfile.dapper (hunk 1 above)
Comment (dimaunx) [reply Dockerfile.dapper]: We can create more layers with separate RUN, but each later is 1.4 GB. We wanted to minimize the number of layers. The alternative is bunch of 1.4 GB layers laying around :).

Comment (dimaunx) [reply Dockerfile.dapper]: IN any case we should host the build image on some public docker registry and pull it only once, this will improve the times considerably without the need to wait for dependencies on every change.

File: Dockerfile.dapper [hunk 7]
```diff
@@ -4,7 +4,8 @@ FROM ubuntu:16.04
 ARG DAPPER_HOST_ARCH
 ENV HOST_ARCH=${DAPPER_HOST_ARCH} ARCH=${DAPPER_HOST_ARCH} DAPPER_ENV=REPO DAPPER_ENV=TAG \
     DAPPER_SOURCE=/go/src/github.com/rancher/submariner DAPPER_DOCKER_SOCKET=true \
-    TRASH_CACHE=${DAPPER_SOURCE}/.trash-cache HOME=${DAPPER_SOURCE} LINT_VERSION=v1.16.0 DAPPER_OUTPUT=output
+    TRASH_CACHE=${DAPPER_SOURCE}/.trash-cache HOME=${DAPPER_SOURCE} DAPPER_OUTPUT=output \
+    LINT_VERSION=v1.17.1 HELM_VERSION=v2.14.1 KIND_VERSION=v0.3.0 KUBEFED_VERSION=0.1.0-rc2
```
Comment (skitt) [inline Dockerfile.dapper]: Could you split these up into separate lines, like #82? It means version bumps produce nice diffs in future...

File: pkg/client/clientset/versioned/clientset.go [hunk 8]
```diff
@@ -28,8 +28,6 @@ import (
 type Interface interface {
 	Discovery() discovery.DiscoveryInterface
 	SubmarinerV1() submarinerv1.SubmarinerV1Interface
-	// Deprecated: please explicitly pick a version if possible.
```
Comment (skitt) [inline pkg/client/clientset/versioned/clientset.go]: This shouldn’t be part of this patch.

File: pkg/client/clientset/versioned/fake/clientset_generated.go [hunk 9]
```diff
@@ -75,8 +75,3 @@ var _ clientset.Interface = &Clientset{}
 func (c *Clientset) SubmarinerV1() submarinerv1.SubmarinerV1Interface {
 	return &fakesubmarinerv1.FakeSubmarinerV1{Fake: &c.Fake}
 }
-
```
Comment (skitt) [inline pkg/client/clientset/versioned/fake/clientset_generated.go]: This shouldn’t be part of this patch either.

File: pkg/controllers/datastoresyncer/datastoresyncer.go [hunk 10]
```diff
@@ -83,7 +83,7 @@ func (d *DatastoreSyncer) ensureExclusiveEndpoint() {
 
 	for _, endpoint := range endpoints {
 		if !util.CompareEndpointSpec(endpoint.Spec, d.localEndpoint.Spec) {
-			endpointCrdName, err := util.GetEndpointCRDName(endpoint)
+			endpointCrdName, err := util.GetEndpointCRDName(&endpoint)
```
Comment (skitt) [inline pkg/controllers/datastoresyncer/datastoresyncer.go]: This also looks like it was pulled in from another patch, is that right?

File: pkg/datastore/datastore.go [hunk 11]
```diff
@@ -14,22 +14,33 @@ import (
 type Datastore interface {
 	// This gets all clusters that match the color code
 	GetClusters(colorCodes []string) ([]types.SubmarinerCluster, error)
+
 	// This gets a single cluster based on cluster ID and returns a v1.Cluster or error based on what it retrieves
-	GetCluster(clusterID string) (types.SubmarinerCluster, error)
```
Comment (skitt) [inline pkg/datastore/datastore.go]: Likewise, this is from another patch.

File: scripts/kind-e2e/e2e.sh [hunk 12]
```diff
@@ -39,7 +39,7 @@ function install_helm() {
             echo Installing helm on cluster${i}.
             kubectl -n kube-system create serviceaccount tiller
             kubectl create clusterrolebinding tiller --clusterrole=cluster-admin --serviceaccount=kube-system:tiller
-            helm init --service-account tiller
+            helm init --service-account tiller > /dev/null 2>&1
```
Comment (skitt) [inline scripts/kind-e2e/e2e.sh]: I know the output here is boring, but IME it’s better to keep the output because stuff ends up going wrong :-(.

File: scripts/kind-e2e/e2e.sh [hunk 13]
```diff
@@ -51,7 +51,7 @@ function setup_broker() {
         echo Submariner CRDs already exist, skipping broker creation...
     else
         echo Installing broker on cluster1.
-        helm install submariner-latest/submariner-k8s-broker --name ${SUBMARINER_BROKER_NS} --namespace ${SUBMARINER_BROKER_NS}
+        helm install submariner-latest/submariner-k8s-broker --name ${SUBMARINER_BROKER_NS} --namespace ${SUBMARINER_BROKER_NS} > /dev/null 2>&1
```
Comment (skitt) [inline scripts/kind-e2e/e2e.sh]: Same here (and in other places further down), please keep the output. (Even if we do end up wanting to get rid of it, it should be a separate patch.)

Comment (dimaunx) [reply scripts/kind-e2e/e2e.sh]: > Same here (and in other places further down), please keep the output. (Even if we do end up wanting to get rid of it, it should be a separate patch.)

Yes i agree

File: scripts/kind-e2e/e2e.sh (hunk 12 above)
Comment (dimaunx) [reply scripts/kind-e2e/e2e.sh]: > I know the output here is boring, but IME it’s better to keep the output because stuff ends up going wrong :-(.

Yes i know what you mean, the outputs are way too long :), i have added echo to each function but you are correct we probably should keep the output.

File: scripts/kind-e2e/nginx-federated.sh (hunk 3 above)
Comment (dimaunx) [reply scripts/kind-e2e/nginx-federated.sh]: Fixed, thank you.

File: pkg/client/clientset/versioned/clientset.go (hunk 8 above)
Comment (dimaunx) [reply pkg/client/clientset/versioned/clientset.go]: Sorry for the noise, this should not be here.

File: pkg/client/clientset/versioned/fake/clientset_generated.go (hunk 9 above)
Comment (dimaunx) [reply pkg/client/clientset/versioned/fake/clientset_generated.go]: Sorry for the noise, this should not be here.

File: pkg/controllers/datastoresyncer/datastoresyncer.go (hunk 10 above)
Comment (dimaunx) [reply pkg/controllers/datastoresyncer/datastoresyncer.go]: Sorry for the noise, this should not be here, bad rebase.

File: pkg/datastore/datastore.go (hunk 11 above)
Comment (dimaunx) [reply pkg/datastore/datastore.go]: Sorry for the noise, this should not be here, bad rebase.

File: Dockerfile.dapper (hunk 7 above)
Comment (dimaunx) [reply Dockerfile.dapper]: Done with the latest commit.

File: Dockerfile.dapper (hunk 1 above)
Comment (mangelajo) [reply Dockerfile.dapper]: 1.4GB! :-o why that much? :D

File: scripts/kind-e2e/e2e.sh (hunk 2 above)
Comment (mangelajo) [reply scripts/kind-e2e/e2e.sh]: ack then!

File: Dockerfile.dapper (hunk 1 above)
Comment (dimaunx) [reply Dockerfile.dapper]: apt-get update

File: Dockerfile.dapper [hunk 14]
```diff
@@ -4,7 +4,11 @@ FROM ubuntu:16.04
 ARG DAPPER_HOST_ARCH
 ENV HOST_ARCH=${DAPPER_HOST_ARCH} ARCH=${DAPPER_HOST_ARCH} DAPPER_ENV=REPO DAPPER_ENV=TAG \
     DAPPER_SOURCE=/go/src/github.com/rancher/submariner DAPPER_DOCKER_SOCKET=true \
-    TRASH_CACHE=${DAPPER_SOURCE}/.trash-cache HOME=${DAPPER_SOURCE} LINT_VERSION=v1.16.0 DAPPER_OUTPUT=output
+    TRASH_CACHE=${DAPPER_SOURCE}/.trash-cache HOME=${DAPPER_SOURCE} DAPPER_OUTPUT=output \
+    LINT_VERSION=v1.17.1 \
+    HELM_VERSION=v2.14.1 \
```
Comment (mangelajo) [inline Dockerfile.dapper]: You're still bumping the helm version (I believe skitt was asking to separate on a different PR).

File: scripts/kind-e2e/README.md [hunk 15]
```diff
@@ -103,6 +103,23 @@ the following lucene query example can be used to query the logs.
 kubernetes.namespace:"submariner" AND kubernetes.node.name: (cluster2* OR cluster3*) AND kubernetes.labels.app: submariner*
 ```
 
+#### Federation V2
+Providing **kubefed=true** parameter to **make e2e** command will setup federation.
+Federation control plain will be created on cluster1 and clusters 2/3 will be added as members.
```
Comment (mangelajo) [inline scripts/kind-e2e/README.md]: plain=plane

File: Dockerfile.dapper (hunk 14 above)
Comment (dimaunx) [reply Dockerfile.dapper]: > You're still bumping the helm version (I believe skitt was asking to separate on a different PR).

I think he requested the environment variables for versions to be on separate lines.

Comment (skitt) [reply Dockerfile.dapper]: > > You're still bumping the helm version (I believe skitt was asking to separate on a different PR).
> 
> I think he requested the environment variables for versions to be on separate lines.

Yes, but I also like changes to be isolated ;-). This PR should only be about kubefed integration, and version bumps unrelated to that should be separate IMO. If the kubefed integration part requires version bumps, that should be documented in the commit message, ideally with a link to the changelog for the new version (see #82 for Helm).

Comment (dimaunx) [reply Dockerfile.dapper]: Yes you are right i have pushed too many changes. Will revert the versions and open a separate PR.

File: scripts/kind-e2e/nginx-federated.sh (hunk 4 above)
Comment (dimaunx) [reply scripts/kind-e2e/nginx-federated.sh]: Done

File: scripts/kind-e2e/nginx-federated.sh (hunk 5 above)
Comment (dimaunx) [reply scripts/kind-e2e/nginx-federated.sh]: Done

File: scripts/kind-e2e/nginx-federated.sh (hunk 6 above)
Comment (dimaunx) [reply scripts/kind-e2e/nginx-federated.sh]: Done

File: scripts/kind-e2e/e2e.sh (hunk 13 above)
Comment (dimaunx) [reply scripts/kind-e2e/e2e.sh]: Done with latest commit.

File: Dockerfile.dapper (hunk 14 above)
Comment (dimaunx) [reply Dockerfile.dapper]: Done!

File: Dockerfile.dapper (hunk 1 above)
Comment (dimaunx) [reply Dockerfile.dapper]: @mangelajo After some testing you are right, better solution would be separate each install step in to its own RUN command. Storage is less expensive then time :).

Comment (mangelajo) [reply Dockerfile.dapper]: Thanks!! :)
